### PR TITLE
WIP: keadm getdb : Get and format the data in the output database edgecore.db

### LINKED ISSUE #1918: keadm get : Get and format the data in the output database edgecore.db
**What would you like to be added/modified**:
Get and format the data in the output database edgecore.db like `kubectl get`, help debug.
**Why is this needed**:
Various types of data now are stored in edge-side database **edgecore.db**(/var/lib/kubeedge/edgecore.db)  such as :
- pod
- secret
- configmap
- service
- endpoints
- node

Now, If we want to confirm whether a specific resource is successfully delivered to the edge, we need to use `sqlite3 /var/lib/kubeedge/edgecore.db` to get the data in edgecore.db, as below:
image

It is troublesome and messy comparing to `kubectl get`:
image
image

Tips:
1. How to get data in edgecore.db?
Refer to `edge\pkg\metamanager\dao\meta.go`

2. How to add a command to keadm?
Refer to `keadm\cmd\keadm\app\cmd\edge\join.go`

3. How to fomat data?
Refer to `kubectl get`

### PR DESCRIPTION
**Pull request by Weaverzhu**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind feature

**What this PR does / why we need it**:
keadm getdb : Get and format the data in the output database edgecore.db
**Which issue(s) this PR fixes**:

Fixes #1918

**Special notes for your reviewer**:
Work in progress
**Does this PR introduce a user-facing change?**:

```release-note
examples are shown below, currently support format output of json, yaml is working in progress, 

use "-p=${path to edgecore.db}" to assign the path of edgecore.db

------------------- split -----------------

root@master:~# kubectl get pods -owide
NAME                               READY   STATUS    RESTARTS   AGE   IP            NODE   NOMINATED NODE   READINESS GATES
nginx-deployment-55cfcf74d-qfnwd   1/1     Running   0          25s   10.244.1.20   node   <none>           <none>
nginx-deployment-55cfcf74d-zzn4r   1/1     Running   0          25s   172.17.0.2    edge   <none>           <none>

root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm getdb
KEY
kube-system/pod/kube-flannel-ds-amd64-jxjlg
default/pod/nginx-deployment-55cfcf74d-zzn4r


root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm getdb -o json
---------------kube-system/pod/kube-flannel-ds-amd64-jxjlg---------------
{
        "metadata": {
                "creationTimestamp": "2020-07-16T10:43:37Z",
                "generateName": "kube-flannel-ds-amd64-",
                "labels": {
                        "app": "flannel",
                        "controller-revision-hash": "6fbd6f4c89",
....... // much content omitted

               "podIPs": [
                        {
                                "ip": "192.168.125.12"
                        }
                ],
                "qosClass": "Burstable",
                "startTime": "2020-07-16T10:43:48Z"
        }
}
---------------default/pod/nginx-deployment-55cfcf74d-zzn4r---------------
{
        "metadata": {
                "creationTimestamp": "2020-07-16T12:22:52Z",
                "generateName": "nginx-deployment-55cfcf74d-",
                "labels": {
                        "app": "nginx",
                        "pod-template-hash": "55cfcf74d"
                },
....... // much content omitted
```

### CODE REVIEW

**Comment by Weaverzhu** [comment]
Update: add yaml output

/cc @ls889和 @gsssc
```
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm getdb
KEY
kube-system/pod/kube-flannel-ds-amd64-jxjlg
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm getdb -p ~/dummy.db 
KEY
default/pod/nginx-deployment-55cfcf74d-zzn4r
kube-system/pod/kube-flannel-ds-amd64-jxjlg
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm getdb -oyaml
---------------kube-system/pod/kube-flannel-ds-amd64-jxjlg---------------
metadata:
  creationTimestamp: "2020-07-16T10:43:37Z"
  generateName: kube-flannel-ds-amd64-
  labels:
    app: flannel
    controller-revision-hash: 6fbd6f4c89
    pod-template-generation: "1"
    tier: node
  name: kube-flannel-ds-amd64-jxjlg
  namespace: kube-system
  ownerReferences:
// much content omitted
```

**Comment by GsssC** [comment]
: ), thanx for contributing. We have already a propsal about `keadm debug` in #1939, can you try to fit our command design?

**Comment by Weaverzhu** [comment]
> : ), thanx for contributing. We have already a propsal about `keadm debug` in #1939, can you try to fit our command design?

Ok, thx for guiding, I'll keep an eye on it

**Comment by qingchen1203** [comment]
: ), thanx for contributing. we have a plan to do keadm debug related content #1939, and your PR is consistent with us. Can you try to fit our command design?  Sincerely welcome you to join us.

**Comment by Weaverzhu** [comment]
Update： now I use `corev1.List` to save all infos about resources like kubectl does, and json output is now the same as kubectl does, but yaml output is slightly different which is a bit annoying. Most of the implementation is copied from kubectl, but I'm using `yaml.Marshal` and `json.Marsal` for final output which is different  from kubectl. I'm currently trying to figure it out.

On the other hand, `kubeadm debug get <resourceType>` receive argument as `type` for querying db. So commands `./keadm debug get endpoints` or `./keadm debug get service` should get output as well.

Finally, as for namespaces and selectors, I found implementation in kubectl is a little bit heavy and I'm not sure if I can implement it in a short time. Any guide or help is welcomed!

Difference between `kubectl get pods -ojson` and `./keadm get pod`

```
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm debug get pod -ojson
{
        "apiVersion": "v1",
        "items": [
                {
                        "apiVersion": "v1",
                        "kind": "pod",
                        "metadata": {
                                "creationTimestamp": "2020-07-16T10:43:37Z",
                                "generateName": "kube-flannel-ds-amd64-",
                                "labels": {
                                        "app": "flannel",
                                        "controller-revision-hash": "6fbd6f4c89",

------------------split---------------


root@master:~# kubectl get pods -ojson
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "v1",
            "kind": "Pod",
            "metadata": {
                "creationTimestamp": "2020-07-17T06:43:52Z",
                "generateName": "nginx-deployment-55cfcf74d-",
                "labels": {
                    "app": "nginx",
                    "pod-template-hash": "55cfcf74d"
                },
                "name": "nginx-deployment-55cfcf74d-rts8d",
                "namespace": "default",
                "ownerReferences": [
                    {
                        "apiVersion": "apps/v1",
```

current yaml output:

the term `object` shouldn't be in yaml output, which is different from kubectl, I currently trying to figure it out

```
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm debug get pod -oyaml
object:
  apiVersion: v1
  kind: List
  metadata:
    resourceVersion: ""
    selfLink: ""
items:
- object:
    apiVersion: v1
    kind: pod
    metadata:
      creationTimestamp: "2020-07-16T10:43:37Z"
      generateName: kube-flannel-ds-amd64-
      labels:
        app: flannel
        controller-revision-hash: 6fbd6f4c89
        pod-template-generation: "1"
        tier: node
      name: kube-flannel-ds-amd64-jxjlg
      namespace: kube-system
      ownerReferences:
      - apiVersion: apps/v1
        blockOwnerDeletion: true
```

File: keadm/cmd/keadm/app/cmd/debug/get.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+				"configmap": true,
+				"endpoint":  true,
+			}
+
+			initDb(getDbPath(cmd))
+			if len(args) != 1 {
+				klog.Fatal("need to specify exactly one type of output, e.g: keadm debug get pod")
+			}
+			resourceType := args[0]
+			if !availableResourceTypes[resourceType] {
```
[the remark this answers is not in the document]
Comment (Weaverzhu) [reply keadm/cmd/keadm/app/cmd/debug/get.go:50]: I am currently reading code in kubectl to see if I can reuse part of its code. The implementation in get command is a bit complicated. To use printers in kubectl we need `RESTMapping`. I am trying to figure it out.

File: keadm/cmd/keadm/app/cmd/debug/get.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+				klog.Fatalf("resource type %s is not available", resourceType)
+			}
+
+			namespace, err := getNamespaceFromFlag(cmd)
+			if err != nil {
+				return err
+			}
+
+			result, err := getResult(resourceType)
```
[the remark this answers is not in the document]
Comment (Weaverzhu) [reply keadm/cmd/keadm/app/cmd/debug/get.go:59]: Thx for pointing out, I am planning to read input from args in the same function.

File: keadm/cmd/keadm/app/cmd/debug/get.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (GsssC) [reply keadm/cmd/keadm/app/cmd/debug/get.go:50]: Yes, I also think how to reuse printers matters

**Comment by Weaverzhu** [comment]
`\kubernetes\staging\src\k8s.io\api\core\v1\types.go` contains a lot of useful and interesting struct defitnition and can be used for json parsing. I had some implementation trying to create my own definition, but soon I found it too time-consuming. I guess `RESTMapping` is something maps `runtime.Object` to the structs defined in `types.go`. I'll keep figuring it out.

**Pull request closed without merging**